MSC2666: Get rooms in common with another user (#2666)
matrix-org/matrix-doc: proposals/2666-get-rooms-in-common.md

so they often only have a subset of state for the rooms the user is in. Therefore, the homeserver
should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another
user.

## Proposal

Homeservers will implement a new endpoint `/_matrix/client/v1/user/mutual_rooms`.

This endpoint will take a query parameter of `user_id` which will contain the MXID of the user
matched against.

This endpoint can be rate limited.

The response format will be an array containing all rooms where both the authenticated user and
`user_id` have a membership of type `join`. If the `user_id` does not exist, or does not share any
rooms with the authenticated user, an empty array should be returned.

```http
GET /_matrix/client/v1/user/mutual_rooms/?user_id=%40bob%3Aexample.com
```

```json
{
  "joined": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

The server may decide that the response to this endpoint is too large, and thus an optional key

> `"next_batch_token"` can be inserted, which the client has to pass to `batch_token` in the query

Reviewer: The synapse implementation seems to return a particular error for unknown batch tokens?

Reviewer 2: ...and merged, this thread can be resolved.

it's computationally expensive for both the homeserver and the client, as well as a bandwidth waste for constrained
clients.

## Forward-compatibility considerations

There possibly comes a time where it's desirable to query mutual rooms for more-than-one other user,
where multiple people (including the self-user) are matched against for which rooms all of them
share.

Because of that, the endpoint accepts a query parameter, however, it will only accept *one* query
parameter for the time being. In the future this restriction can be lifted to accept multiple query
parameters under `user_id`

## Security considerations

The information provided in this endpoint is already accessible to the client if it has a copy of all
state that the user can see. This endpoint only makes it possible to get this information without having
to request all state ahead of time.

## Unstable prefix

The implementation MUST use `/_matrix/client/unstable/uk.half-shot.msc2666/user/mutual_rooms`.

The /versions endpoint MUST include a new key in `unstable_features` with the name
`uk.half-shot.msc2666.query_mutual_rooms`.

Previous iterations of this MSC has used the following `unstable_features` key(s):

> - `uk.half-shot.msc2666.mutual_rooms`

Reviewer 2: 🐑 Small addition, as I forgot this a few times prior:

```suggestion
- `uk.half-shot.msc2666.mutual_rooms`
- `uk.half-shot.msc2666`
```

There are a variety of applications for this information. Some users may want to block invites from
users they do not share a room with at the client level, and need a way to poll the homeserver for
this information. Another use case would be trying to determine how a user came across your MXID, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,
modern clients often implement
[lazy-loading of room members](https://spec.matrix.org/v1.3/client-server-api/#lazy-loading-room-members),
so they often only have a subset of state for the rooms the user is in. Therefore, the homeserver
should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another
user.

## Proposal

Homeservers will implement a new endpoint `/_matrix/client/v1/user/mutual_rooms`.

This endpoint will take a query parameter of `user_id` which will contain the MXID of the user

> matched against.

Reviewer 3: If `user_id` can't be specified multiple times, it might belong in the HTTP path instead. Also, can it be specified multiple times?

Reviewer 3: ah, it's listed later in the MSC. Is there a technical reason preventing multiple users from being searched? It feels like an awkward thing to make a future MSC for when we're already here.

Reviewer 2: The discussion context is here: https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r936483418

Reviewer 2: There is no technical reason preventing multiple users for being searched, but I don't see a convincing client usecase, while I agreed to keep the path open for one if the time came, to make it easily extendible.

Reviewer 4: I think if we're *not* putting the user id in the path, then we shouldn't use the `/_matrix/client/v1/user` hierarchy, since most of the endpoints under that hierarchy do have a user_id as the next path component.

Reviewer 2: I'm also realising that if we want to be able to get rooms in common with more users, there could just be another version of this endpoint.

Reviewer 5: With the current MSC text, this seems to be resolved?

Reviewer 2: I want some clear direction on this. The current version (that I merge in this branch via #4035) changes some labels around enough that i'd have to give it another unstable_features tag.

I haven't yet, as i want to sweep the result of this in that tag, before i propose it FCP again.

---

Should I change this back to a path component, or should I keep this as a query element endpoint? I can do either.

Author: This seems to be the stalling factor in this MSC now, and it seems like a fairly trivial matter of just:

 - Drop `/user/`, path becomes `_matrix/client/v1/mutual_rooms`
 - Keep the query parameter, leave the door open to allowing multiple in the future if the need arises.

If everyone is happy with that, we can make the change and re-request FCP.

Reviewer 6: `/v1/mutual_rooms` rather than /v3, but yeah that sounds good to me. No need to define a new unstable prefix either, just change the stable one

> This endpoint can be rate limited.

Reviewer 3: boilerplate:
```suggestion
This endpoint can be rate limited and requires authentication.
```

Reviewer 2: /res #4035

The response format will be an array containing all rooms where both the authenticated user and
`user_id` have a membership of type `join`. If the `user_id` does not exist, or does not share any

> rooms with the authenticated user, an empty array should be returned.

Reviewer 4: is the behaviour on invalid user IDs left as an implementation decision? (I think that's fine if so, but best to say so explicitly)

Reviewer 2: IIRC that would fall under some `400` `Bad Request` semantics, implementation-specific, yeah.

```http

> GET /_matrix/client/v1/user/mutual_rooms/?user_id=%40bob%3Aexample.com

Reviewer 4: this example includes a trailing `/`, which is inconsistent with the definition above.

In general Matrix does not allow extra `/`s in endpoints.

Reviewer 2: Ah, that's an error from a previous change that moved path-element parameters to query parameters.

Reviewer 2: /res https://github.com/matrix-org/matrix-spec-proposals/pull/4035

```

```json
{
  "joined": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

The server may decide that the response to this endpoint is too large, and thus an optional key
`"next_batch_token"` can be inserted, which the client has to pass to `batch_token` in the query
parameters together with the original `user_id` to fetch the next batch of responses. This will

> continue until the server does no longer insert `"next_batch_token"`.

Reviewer 3: Somewhere we need the words "this batch_token is only valid for use with this endpoint", and the semantics for expiration (does the homeserver need to remember the batch token forever?)

Reviewer 2: I think an expiration time of 10 minutes could be reasonable here.

Reviewer 7: In which case we should specify an error case for clients supplying an unknown / expired batch token, so that they can retry.

We can't use `M_UNKNOWN_TOKEN` - as this error would also be returned if the user used an unknown access/refresh token to access this endpoint.

We may need to define a new `errcode` - perhaps `M_UNKNOWN_PAGINATION_TOKEN`?

Reviewer 2: #4035

I have added the 10 minute timer, and added text that it will handle expired tokens the same as invalid tokens, so that implementations may forget those tokens entirely. Do you think this is a good method?

Then the clients' default response to "invalid token" is just to start over, without any token.

The only problem i see is that clients may loop on a broken implementation, where it is not passing the right token back to the server will cause it to get stuck in its mechanism where it continually retries.

Reviewer 7: I would just leave the validity period up to the implementation, but *recommend* at least 10 minutes. Clients should be prepared to handle tokens that do have a ridiculously short window though.

I would also recommend clients limit the maximum number of retries before giving up, precisely to prevent infinite loops as you describe.

```json5
{
  "joined": [
    // ...
  ],

>   "next_batch_token": "<any valid ascii string up to 32 chars>"

Reviewer 2: I'm just now realising that this might cause problems, and we should restrict this to just `[a-zA-Z0-9]` to prevent issues with url encoding.

No implementation (right now) generates a batch token, so if the below suggestion is accepted, no implementations would have to be altered.

```suggestion
  "next_batch_token": "<a string up to 32 characters abiding by the regex /[a-zA-Z0-9]+/>"
```

Reviewer 8: It may be convenient to also include `-_` so that data can be URL-safe base64 encoded. This provides a bit of an escape hatch for encoding arbitrary data. This way 24 binary bytes can be encoded into the token. The token should of course be considered opaque to the client still.

>   "next_batch_token": "<a string up to 32 characters abiding by the regex /[a-zA-Z0-9]+/>"

Reviewer 7: Other spots in the spec call a similar field `next_batch` instead of `next_batch_token`. It seems confusing to deviate?

* https://spec.matrix.org/v1.7/client-server-api/#post_matrixclientv3search
* https://spec.matrix.org/v1.7/client-server-api/#pagination
* https://spec.matrix.org/v1.7/client-server-api/#get_matrixclientv1roomsroomidthreads
* etc

Reviewer 2: Apologies, I wasn't aware of the naming convention.

According to this, its both `next_batch` in the response, and `next_batch` in the query request.

I could update, but that would be a bunch more churn. If you like to see me do it, just say it.

Reviewer 7: Implementations will need to update to remove unstable prefixes anyways, so why not do a parameter rename at the same time?

The implementations don't need to have the exact names to be considered proven and unblock FCP.

Reviewer 3: Pagination is also specified as https://spec.matrix.org/v1.7/appendices/#pagination ftr

Reviewer 2: /res #4035

>   "next_batch_token": "<a string up to 32 characters abiding by the regex /[a-zA-Z0-9]+/>"

Reviewer 2: Per [this comment](https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r1218010122)

```suggestion
  "next_batch_token": "<a string up to 32 characters abiding by the regex /[a-zA-Z0-9_-]+/>"
```

Reviewer 3: can we just use https://github.com/matrix-org/matrix-spec-proposals/blob/rav/proposals/id_grammar/proposals/1597-id-grammar.md#opaque-ids please?

Ideally by name: "a batch_token is an opaque identifier, currently capable of containing the characters [0-9a-zA-Z._~-], must be at most 255 characters, and non-empty"

Reviewer 2: Thanks, I was looking for something like that, will add it 👍

Reviewer 2: /res #4035

}
```

The response error for when the given `batch_token` is invalid will be a response with HTTP code 400,
with `M_INVALID_PARAM` as `errcode`.

The response error for trying to get shared rooms with yourself will be an HTTP code 422, with

> `M_INVALID_PARAM` as the `errcode`.

Reviewer 4: Not a strong opinion here, but 422 is pretty obscure (and we don't use it anywhere else in matrix, afaik).

The *idea* in Matrix is that we shouldn't rely too heavily on HTTP response codes, to make it easier (one day) to move to alternative transports. So if we need to distinguish this from other error cases, we'd do that via a custom `errcode` rather than HTTP response code.

Reviewer 5: Ftr, [CoAP](https://www.rfc-editor.org/rfc/rfc7252#section-5.9) doesn't have code 4.22 or anything semantically similar to HTTP's 422, and there don't seem to be subsequent RFCs addressing that (while 429, e.g., has got is CoAP equivalent in a [later RFC](https://datatracker.ietf.org/doc/html/rfc8516)).

Reviewer 7: Doing this sounds like programmer error vs. something that would actually need to be explicitly handled in practice. So this could just be a `400` / `M_UNKNOWN` - "you cannot request rooms in common with yourself".

Reviewer 2: I've changed the error to 400 + M_UNKNOWN

/res #4035

There possibly comes a time where it's desirable to query mutual rooms for more-than-one other user,
where multiple people (including the self-user) are matched against for which rooms all of them
share.

Because of that, the endpoint accepts a query parameter, however, it will only accept *one* query
parameter for the time being. In the future this restriction can be lifted to accept multiple query
parameters under `user_id`

## Security considerations

The information provided in this endpoint is already accessible to the client if it has a copy of all
state that the user can see. This endpoint only makes it possible to get this information without having
to request all state ahead of time.

## Unstable prefix

The implementation MUST use `/_matrix/client/unstable/uk.half-shot.msc2666/user/mutual_rooms`.

The /versions endpoint MUST include a new key in `unstable_features` with the name
`uk.half-shot.msc2666.query_mutual_rooms`.

Previous iterations of this MSC has used the following `unstable_features` key(s):
- `uk.half-shot.msc2666.mutual_rooms`
- `uk.half-shot.msc2666`

If the value is false or the key is not present, clients MUST assume the feature is not available.

> Once the MSC has been merged, and the homeserver has advertised support for the Matrix version that
> this endpoint is included in, clients should use `/_matrix/client/v1/user/mutual_rooms` and will no
> longer need to check for the `unstable_features` flag.

Reviewer 3: ```suggestion
Once this MSC is considered stable, implementations can use the stable endpoint instead.
```

It may also be worth reading https://github.com/matrix-org/matrix-spec-proposals/pull/4024 to match the language there, as this seems like the sort of feature a client would want to use between FCP completing and spec release.

Reviewer 2: Ah, I think with this original language I remember/wrote this down to "consider stable" to be "when it is included in a spec version"

Together with richvdh's comment [here](https://github.com/matrix-org/matrix-spec-proposals/pull/4024#discussion_r1219377897), i think that for this MSC it is unnecessary to introduce a `.stable` `unstable_features` flag.

Reviewer 2: /res

Reviewer 7: @tulir I assume you (and other clients) will want to use this before homeserver implementations get around to advertising support for the Matrix version this will merge into?

Reviewer 6: Oh yes, stable flags are needed for basically all endpoints

matched against.

This endpoint can be rate limited and requires authentication.
Guests are not allowed to use the endpoint, as guests will usually not join multiple rooms.

The response format will be an array containing all rooms where both the authenticated user and
`user_id` have a membership of type `join`.

If the `user_id` does not exist, or does not share any rooms with the authenticated user,
an empty array should be returned.

Handling invalid user IDs should result in an error, is likely implementation-specific,
and is beyond the scope of this proposal.

```http
GET /_matrix/client/v1/user/mutual_rooms?user_id=%40bob%3Aexample.com
```

```json
{
  "joined": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

The server may decide that the response to this endpoint is too large, and thus an optional key
`"next_batch"` can be inserted, which the client has to pass to `from` in the query
parameters together with the original `user_id` to fetch the next batch of responses. This will
continue until the server does no longer insert `"next_batch"`. The `next_batch` field follows

> the [standard for opaque identifiers](https://spec.matrix.org/v1.16/appendices/#opaque-identifiers).

Reviewer 6: Actually, if the reason for adding pagination is primarily response size, then I can think of a relatively trivial implementation:

1. compute entire response list
2. sort list lexicographically
3. cut off response at a certain index
4. return the last room id as the next batch token
5. when a request has a from token, drop items that are lexicographically lower than the token before cutting the response list

> the [standard for opaque identifiers](https://spec.matrix.org/v1.16/appendices/#opaque-identifiers).

Reviewer 6: Only problem with that is the character set for opaque identifiers doesn't allow `!`, plus pre-v12 room IDs can have arbitrary unicode. Maybe it can be base64 encoded and truncated? Just need handling for the silly user case of having more than the batch size of pre-v12 rooms that have the same first 191 letters (e.g. don't return next batch token if new token matches old one)

matched against.

This endpoint can be rate limited and requires authentication.
Guests are not allowed to use the endpoint, as guests will usually not join multiple rooms.

The response format will be an array containing all rooms where both the authenticated user and
`user_id` have a membership of type `join`.

If the `user_id` does not exist, or does not share any rooms with the authenticated user,
an empty array should be returned.

Handling invalid user IDs may result in an error, is likely implementation-specific,
and is beyond the scope of this proposal. Implementations are also free to simply
return a success response with an empty array for invalid user IDs.

```http
GET /_matrix/client/v1/user/mutual_rooms?user_id=%40bob%3Aexample.com
```

```json
{
  "joined": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

The server may decide that the response to this endpoint is too large, and thus an optional key
`"next_batch"` can be inserted, which the client has to pass to `from` in the query
parameters together with the original `user_id` to fetch the next batch of responses. This will

> continue until the server does no longer insert `"next_batch"`. The `next_batch` field follows

Reviewer 7: ```suggestion
continue until the server no longer inserts `"next_batch"`. The `next_batch` field follows
```

the [standard for opaque identifiers](https://spec.matrix.org/v1.16/appendices/#opaque-identifiers).

```json5
{
  "joined": [
    // ...
  ],
  "next_batch": "<an opaque identifier, containing only the characters [0-9a-zA-Z._~-], non-empty if not omitted, and at most 255 characters>"
}
```

The batch tokens this endpoint generates are only valid for this endpoint.

The response error for when the given `from` batch token is invalid will be a response with HTTP code 400,
with `M_INVALID_PARAM` as `errcode`.

The response error for trying to get shared rooms with yourself will be an HTTP code 400, with
`M_UNKNOWN` as the `errcode`. And the error description may be "you cannot request rooms in common with yourself".

> Tokens generated by this endpoint must be valid for at least 10 minutes, after which, tokens can expire.

Reviewer 7: ```suggestion
Batch tokens generated by this endpoint SHOULD be valid for at least 10 minutes, after which, tokens can expire.
```

It's impossible to enforce this in [the Synapse implementation](https://github.com/element-hq/synapse/pull/19279), which just uses the base64url-encoded, last-returned Room ID as the batch token. It will be invalidated by the user joining/leaving rooms or rooms being purged by the homeserver.

an empty array should be returned.

Handling invalid user IDs may result in an error, is likely implementation-specific,
and is beyond the scope of this proposal. Implementations are also free to simply
return a success response with an empty array for invalid user IDs.

```http
GET /_matrix/client/v1/user/mutual_rooms?user_id=%40bob%3Aexample.com
```

```json
{
  "joined": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

The server may decide that the response to this endpoint is too large, and thus an optional key
`"next_batch"` can be inserted, which the client has to pass to `from` in the query
parameters together with the original `user_id` to fetch the next batch of responses. This will
continue until the server no longer inserts `"next_batch"`. The `next_batch` field follows
the [standard for opaque identifiers](https://spec.matrix.org/v1.16/appendices/#opaque-identifiers).

```json5
{
  "joined": [
    // ...
  ],
  "next_batch": "<an opaque identifier, containing only the characters [0-9a-zA-Z._~-], non-empty if not omitted, and at most 255 characters>"
}
```

> 

Reviewer 7: ```suggestion

The client can make another request to get the next batch of results:

```http
GET /_matrix/client/v1/user/mutual_rooms?user_id=%40bob%3Aexample.com&from=<opaque_identifier>
```
```

(Note: GitHub won't render the above properly due to the additional set of backticks.)

The batch tokens this endpoint generates are only valid for this endpoint.

The response error for when the given `from` batch token is invalid will be a response with HTTP code 400,

> with `M_INVALID_PARAM` as `errcode`.

Reviewer 7: ```suggestion
If the batch token provided by the client is deemed invalid by the server, the server
MAY return the standard error response with HTTP code 400 and `errcode`
`M_INVALID_PARAM`.
```

stores membership information for each of its users, the information may not be stored in a way that
is readily accessible. Homeservers that have implemented
[POST /user_directory/search](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3user_directorysearch)
may have started some of this work, if they are limiting users to searching for users for which they
share rooms. While this is not a given by any means, it may mean that implementations of this API
and /search may be complimentary.

## Alternatives

A client which holds full and current state can already see all membership for all rooms, and thus
determine which of those rooms contains a "join" membership for the given user_id. Clients which "lazy-load"
however do not have this information, as they will have only synced a subset of the full membership for
all rooms. While a client *could* pull all membership for all rooms at the point of needing this information,
it's computationally expensive for both the homeserver and the client, as well as a bandwidth waste for constrained
clients.

## Forward-compatibility considerations

> There possibly comes a time where it's desirable to query mutual rooms for more-than-one other user,

Reviewer 7: ```suggestion
There possibly comes a time where it's desirable to query mutual rooms for more than one other user,
```

where multiple people (including the self-user) are matched against for which rooms all of them
share.

Because of that, the endpoint accepts a query parameter, however, it will only accept *one* query
parameter for the time being. In the future this restriction can be lifted to accept multiple query

> parameters under `user_id`

Reviewer 7: ```suggestion
parameter for the time being. This restriction may be lifted by a future MSC.
```

determine what they have in common before participating in conversation.

There are a variety of applications for this information. Some users may want to block invites from
users they do not share a room with at the client level, and need a way to poll the homeserver for
this information. Another use case would be trying to determine how a user came across your MXID, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,
modern clients often implement
[lazy-loading of room members](https://spec.matrix.org/v1.3/client-server-api/#lazy-loading-room-members),
so they often only have a subset of state for the rooms the user is in. Therefore, the homeserver
should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another
user.

## Proposal

Homeservers will implement a new endpoint `/_matrix/client/v1/mutual_rooms`.

> This endpoint will take a query parameter of `user_id` which will contain the MXID of the user

Reviewer 7: ```suggestion
This endpoint will take a required query parameter of `user_id` which will contain the MXID of the user
```

matched against.

This endpoint can be rate limited and requires authentication.
Guests are not allowed to use the endpoint, as guests will usually not join multiple rooms.

The response format will be an array containing all rooms where both the authenticated user and
`user_id` have a membership of type `join`.

If the `user_id` does not exist, or does not share any rooms with the authenticated user,
an empty array should be returned.

Handling invalid user IDs may result in an error, is likely implementation-specific,
and is beyond the scope of this proposal. Implementations are also free to simply

> return a success response with an empty array for invalid user IDs.

Reviewer: This paragraph is a bit weird. I'd suggest making it an error instead:

```suggestion
If an invalid user ID is provided the server MUST respond with a 400 error with
code `M_INVALID_PARAM`.
```

Reviewer 6: Currently the MSC says to leave it as an implementation detail rather than encoding anything into the spec. Other endpoints like `GET /profile/{userId}` also don't specify any particular handling of invalid user IDs. I'd prefer doing it consistently for all endpoints rather than just this one.

Reviewer 9: @mscbot concern should an error be required for invalid user IDs

Reviewer: Then let's just spec something sane as an edge case instead of leaving it up to the implementation.

```http
GET /_matrix/client/v1/user/mutual_rooms?user_id=%40bob%3Aexample.com
```

```json
{
  "joined": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

The server may decide that the response to this endpoint is too large, and only return
a subset of the results. In this case, the server should populate the optional field
`"next_batch"` with an

> [opaque identifier](https://spec.matrix.org/v1.16/appendices/#opaque-identifiers).

Reviewer 6: In the current implementations adding the count would be trivial. It might not be in future implementations, but maybe that doesn't matter? The primary goal of pagination is ensuring the server doesn't have to send arbitrarily large response payloads.

We could define that the count may be inaccurate or something like that to let servers skip calculating the exact count if they feel it's too difficult

what they have in common before participating in converstion.

There are a variety of applications for this information. Some users may want to block invites from
users they do not share a room with at the client level, and need a way to poll the homeserver for
this information. Another use case would be trying to determine how a user came across your mxid, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,
modern clients tend to implement "lazy-loaded" design patterns, so they often only have state for the
rooms the user has interacted with, or at least a subset of all rooms they are in. Therefore, the homeserver
should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another user.

## Proposal

Homeservers should implement a new endpoint `/users/{user_id}/shared_rooms/{other_user_id}` which will take
the current users mxid and the other users mxid, which the user is trying to search for.

> The response format will be an array containing all rooms where both the `user_id` and `other_user_id` have

Reviewer 10: It should be required that one of the user IDs is your own user ID, thus also requiring authentication via access token

Reviewer 6: It would probably make most sense to require `user_id` to be your own user ID

Author: I was riffing off what we do for account_data. My thought process was as follows:

- Seems a shame to add a new root `/shared_rooms` path, I wonder if I can use `/rooms`.
- Hm, no. `/rooms` is for fetching data about a single room. How about `/users`
- Oh, but all `/users` calls expect the owners user_id.

So in the end I could either break the `/users` convention, write a new path, or use the existing convention. I went with C, but I am happy to hear arguments otherwise.

Author: > It would probably make most sense to require `user_id` to be your own user ID

That's what I am doing in the code, perhaps I need to spell it out :P

Reviewer 11: In conduit I fully ignore the user id in the path and always use the user id from the access token. Is that okay?

Author: In practise that's fine. It's a bit of a spec wart that we have the user_id in the path, because the spec does not allow for it to differ from the access_token.

> a membership of type `join`. 

Reviewer 11: Wouldn't it also be useful to return rooms which you both participated in, in the past, but the other user was banned/kicked. So instead of returning a list of rooms you both joined, the endpoint could return multiple lists, one for each membership state

Author: In theory we could stick a ?membership=[types] on the end of this endpoint, which is a comma separated list of allowed membership states for the remote user. I guess we could even do it for the current user too, however I'm unsure how useful this is.

Reviewer 10: This unfortunately leaks information (and thus soru is strongly in favour *not* to do it):

Suppose alice and bob shared a room at once point, but alice left. Via that endpoint she can then keep querying the room if bob is still in there.

Reviewer 11: @Sorunome I meant room where alice is joined, but bob has a different membership state. In the UI it would show all rooms alice joined and bob has *any* membership event in, with some text saying which one it is.

Reviewer 10: Aaaah, that would be ok, as in, that wouldn't leak information, thanks for clarification!

Reviewer 11: Another idea is to also show the power level, sothe Ui can say that a user is moderator, admin or muted in some room

Reviewer 10: That would also require fetching power level events for a given room, not only search `m.room.member`, and thus would (greatly?) increase complexity on the server side for little gain?

Reviewer 11: I don't think getting two events of the current state is expensive, but maybe in synapse's architecture it is. Idk

Author: I think it's important to keep this MSC scoped in order to improve it's chances of ~~not getting bikeshedded~~ keeping the implementation easy for *all* homeservers.

I'm going to consciously prioritise the use case of showing a simple list of rooms that you are both in, and clients can populate their UI with any state they have for the room.

Reviewer 11: Perhaps you can rename the `rooms` field to `join` then, so it's easy to add to it in the future?

Author: Yep, I think that's a fair compromise.

Author: I've gone with "joined" to line up with things like https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-rooms-roomid-joined-members, but otherwise it's been included.

```
GET _matrix/client/unstable/users/@alice:example.com/shared_rooms/@bob:example.com
```

```json

> [

Reviewer 10: Shouldn't the reply be

```json
{
  "rooms": [
    "!blah",
    "!blubb",
  ]
}
```
As so far *no* endpoint in the spec returns an array, only objects?

Author: Yeah, agreed. I was lazy here.

# MSC 2666: Get rooms in common with another user

It is useful to be able to fetch rooms you have in common with another user. Popular messaging services
such as Telegram offer users the ability to show "groups in common", which allows users to determine
what they have in common before participating in converstion.

There are a variety of applications for this information. Some users may want to block invites from
users they do not share a room with at the client level, and need a way to poll the homeserver for
this information. Another use case would be trying to determine how a user came across your mxid, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,

> modern clients often implement "lazy-loadeding", so they often only have state for a subset of all rooms

Reviewer 9: ```suggestion
modern clients often implement "lazy-loading", so they often only have a subset of state for the rooms
```

the user is in. Therefore, the homeserver should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another user.

## Proposal

Homeservers should implement a new endpoint `/users/{current_user}/shared_rooms/{other_user_id}` which will take

> the authenticated users MxID and the user that is being searched for.

Reviewer 9: ```suggestion
the authenticated user's MxID and the user that is being searched for.
```

The response format will be an array containing all rooms where both the `current_user` and `other_user_id` have

> a membership of type `join`. If either user is not joined to any rooms, or the `other_user_id` does not exist, an

Reviewer 11: What happens if current_user == other_user ?

Author: Good point. In my opinion it's an invalid request, as you can just called /joinedRooms for the same result except that path can optimise for the current user.

Reviewer 12: So something like this?

Servers may reject the request, with [some appropriate error code], if `other_user_id` is the same as `current_user`.  Clients should use /joinedRooms to determine what rooms the current user has joined instead.

Author: It's a tossup between consistency and being too strict. I am minded to be lean on a strong requirement to reject, because doing this has no purpose since another endpoint exists.

# MSC 2666: Get rooms in common with another user

It is useful to be able to fetch rooms you have in common with another user. Popular messaging services
such as Telegram offer users the ability to show "groups in common", which allows users to determine
what they have in common before participating in converstion.

There are a variety of applications for this information. Some users may want to block invites from
users they do not share a room with at the client level, and need a way to poll the homeserver for
this information. Another use case would be trying to determine how a user came across your mxid, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,
modern clients often implement "lazy-loading", so they often only have a subset of state for the rooms
the user is in. Therefore, the homeserver should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another user.

## Proposal

> Homeservers should implement a new endpoint `/users/{current_user}/shared_rooms/{other_user_id}` which will take

Reviewer 9: Why `/users/` rather than `/user/` as other existing endpoints use?

Reviewer 6: Seems to be a typo in the MSC, it's implemented as `/user/`

Author: fixed

the authenticated user's MxID and the user that is being searched for.

The response format will be an array containing all rooms where both the `current_user` and `other_user_id` have
a membership of type `join`. If either user is not joined to any rooms, or the `other_user_id` does not exist, an
empty array should be returned. If the `current_user` and `other_user_id` are the same, then the endpoint SHOULD
reject with HTTP 400.

```

> GET _matrix/client/unstable/uk.half-shot.msc2666/users/@alice:example.com/shared_rooms/@bob:example.com

Author: @anoadragon453 asks if we should continue to use the /users/{user_id}/.../{other_user_id}/ format given we don't allow you to do it on behalf of other users. This was probably intended as a way for appservices to puppet their accounts a long time ago before as_user became the way to do things like this

Author: IMO, I'm fine with dropping this OR we could decide that homeservers can decide whether to allow other user_ids so that admins can spy on their users?

Reviewer 7: I was thinking about this functionality, but I can't really visualize how this would be useful in client UI.

I could see something that would allow you to see what rooms an arbitrary number of users share though? That's probably out of scope here though.

Admins can also usually just go and have a poke in the DB as well.

Author: Yeah, I think at best this would be convenience but frankly it means every client now needs to paste in their user_id in order to call the api. That sucks. Let's drop it.

Reviewer 11: I'm for dropping as well. Maybe even replace existing endpoints with non-userid paths (or support both for a while)

> GET _matrix/client/unstable/uk.half-shot.msc2666/users/@alice:example.com/shared_rooms/@bob:example.com

Reviewer 6: Isn't the MSC supposed to use the actual endpoint and not vendor-prefixed ones?

```

```json
{
  "rooms": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

## Potential issues

Homeserver performance and storage may be impacted by this endpoint. While a homeserver already stores
membership information for each of its users, the information may not be stored in a way that is readily
accessible. Homeservers that have implemented [POST /user-directory/search](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-user-directory-search)
may have started some of this work, if they are limiting users to searching for users for which they
share rooms. While this is not a given by any means, it may mean that implementations of this API
and /search may be complimentary.

## Alternatives

A client can already read all membership for all rooms, and thus determine which of those rooms contains
a "join" membership for the given user_id. However, this method is computationally expensive on the homeserver
and the client. Furthermore, it would increase total network traffic (which is important for low bandwith / mobile clients)

> as well as include lots of extranious information.

Reviewer 3: This does not feel like a compelling reason to add an endpoint to the spec. Low bandwidth and mobile clients should be using more restrictive filters to limit their exposure to information, and if a client isn't using a local store for checking this then they've really messed up how this feature could work. Clients which would want this feature would have already cached state locally.

Reviewer 6: I don't really get your point. Do you mean there should be some non-standalone-endpoint way to get the information without loading all room state?

The only way for this to work without any API changes is if clients either stop lazy loading or request the member lists for *all* rooms at the same time. What's wrong with wanting this feature and lazy loading at the same time?

Reviewer 3: Lazy-loading is a much stronger justification than the words here.

Author: Lazy-loading is mentioned [here](https://github.com/matrix-org/matrix-doc/pull/2666/files/db99583ce577582f6bfe4b28d6ecfa7b1bf54590#diff-366ea8b6ee99e7bf4f9501ae6e9ac376R13-R15) and it was meant to be taken in mind when reading the rest of the MSC, but I'll make a point of including it in this section too.

users they do not share a room with at the client level, and need a way to poll the homeserver for
this information. Another use case would be trying to determine how a user came across your mxid, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,
modern clients often implement "lazy-loading", so they often only have a subset of state for the rooms
the user is in. Therefore, the homeserver should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another user.

## Proposal

Homeservers should implement a new endpoint `/user/shared_rooms/{user_id}` which will take
the MxID of the user who is being searched for.

The response format will be an array containing all rooms where both the authenticated user and `user_id` have
a membership of type `join`. If the `user_id` does not exist, or does not share any rooms with the authenticated user,
an empty array should be returned.

```

> GET _matrix/client/r0/user/shared_rooms/@bob:example.com

Reviewer 7: ```suggestion
GET /_matrix/client/r0/user/shared_rooms/%40bob%3Aexample.com
```

```

```json
{
  "joined": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

> 

Reviewer 7: Could you specify some expected error codes here please? Such as the code when trying to request shared rooms with yourself.

Author: I think there is only one error situation likely to occur, which is the one you mentioned. Users who do not exist will return an empty list, as will invalid user_ids.

Reviewer 7: I probably would've said `M_INVALID_PARAM`, but I don't think it matters much.

may have started some of this work, if they are limiting users to searching for users for which they
share rooms. While this is not a given by any means, it may mean that implementations of this API
and /search may be complimentary.

## Alternatives

A client which holds full and current state can already see all membership for all rooms, and thus
determine which of those rooms contains a "join" membership for the given user_id. Clients which "lazy-load"
however do not have this information, as they will have only synced a subset of the full membership for
all rooms. While a client *could* pull all membership for all rooms at the point of needing this information,
it's computationally expensive for both the homeserver and the client, as well as a bandwidth waste for contrained
clients.

## Security considerations

The information provided in this endpoint is already accessible to the client if it has a copy of all
state that the user can see. This endpoint only makes it possible to get this information without having
to request all state ahead of time.

## Unstable prefix

The implementation MUST use `/_matrix/client/unstable/uk.half-shot.msc2666/user/shared_rooms/{user_id}`.  

> The /versions endpoint MUST include a new key in `unstable_features` with the name `uk.half-shot.msc2666`.  

Reviewer 7: Just checking that you're stating that `uk.half-shot.msc2666` has to be present, but could either be `true` or `false` to indicate that this homeserver supports this feature.

Author: Have spelt it out, but yes.

> # MSC 2666: Get rooms in common with another user

Reviewer 2: @ spec scribe;

I've added an implementation which updates the JS SDK code to use the new naming; https://github.com/matrix-org/matrix-js-sdk/pull/2271

It is useful to be able to fetch rooms you have in common with another user. Popular messaging services
such as Telegram offer users the ability to show "groups in common", which allows users to determine
what they have in common before participating in converstion.

There are a variety of applications for this information. Some users may want to block invites from
users they do not share a room with at the client level, and need a way to poll the homeserver for
this information. Another use case would be trying to determine how a user came across your mxid, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,
modern clients often implement "lazy-loading", so they often only have a subset of state for the rooms
the user is in. Therefore, the homeserver should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another user.

## Proposal

> Homeservers should implement a new endpoint `/user/shared_rooms/{user_id}` which will take

Reviewer 8: Nit: `shared_rooms` could be taken to mean either "rooms I have shared with this user" or "rooms that I share with this user". It is probably better to call this `common_rooms`. (Although that may be confused for common as in popular, but that doesn't make sense when passing another user ID).

Reviewer 2: I kinda agree, we could probably rename it to "mutual_rooms", just like discord has "mutual friends/servers".

Reviewer 8: I think that name is quite clear.

Author: Mutual sounds good to me, and probably the most clearest one.

the MxID of the user who is being searched for.

The response format will be an array containing all rooms where both the authenticated user and `user_id` have
a membership of type `join`. If the `user_id` does not exist, or does not share any rooms with the authenticated user,
an empty array should be returned.

```

> GET /_matrix/client/r0/user/mutual_rooms/%40bob%3Aexample.com

Reviewer 2: There should probably be a way to refuse service to this endpoint with no data, see https://github.com/matrix-org/synapse/pull/12038#discussion_r819599640

I suggest responding with a 404 as a known general failure condition, in addition to "user not found".

> GET /_matrix/client/r0/user/mutual_rooms/%40bob%3Aexample.com

Reviewer 7: ```suggestion
GET /_matrix/client/v1/user/mutual_rooms/%40bob%3Aexample.com
```

To follow the new per-endpoint versioning scheme.

```

```json
{
  "joined": [
    "!OGEhHVWSdvArJzumhm:matrix.org",
    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

> The response error for trying to get shared rooms with yourself will be an HTTP code 400, with `M_FORBIDDEN` as the `errcode`.

Reviewer 7: Should this be an HTTP `403` since it's `M_FORBIDDEN`?

Author: Doesn't have to be. [403](https://httpstatuses.com/403) "The server understood the request but refuses to authorize it." would imply an authentication failure which would leave clients confused because they are using correct authentication information.

[400](https://httpstatuses.com/400) "The server cannot or will not process the request due to something that is perceived to be a client error" which would be correct here, since the request cannot be processed due to incorrect parameters. The request is still FORBIDDEN, but not due to authentication. This is the closest approximation I could find in the spec. `M_INVALID_PARAM` would have worked, but the spec implies that this is due to incorrect typing which isn't the case here.

I think this is fine.

Reviewer 13: I've used 422 Unprocessable Entity in some of my own projects.

Author: 422 would actually make the most sense, but I couldn't find any usage of it in Synapse/TheSpec so I assumed it was unpopular (as 410+ error codes often are..)

## Potential issues

Homeserver performance and storage may be impacted by this endpoint. While a homeserver already stores
membership information for each of its users, the information may not be stored in a way that is readily
accessible. Homeservers that have implemented [POST /user-directory/search](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-user-directory-search)
may have started some of this work, if they are limiting users to searching for users for which they
share rooms. While this is not a given by any means, it may mean that implementations of this API
and /search may be complimentary.

> ## Alternatives

Reviewer 8: Does it make sense to extend the search API with member constraints?

This would be more generic and be able to specify multiple usersto include? This may be helpful for finding old rooms (I remember that I was walking with Alice and Bob about this but can't find the room).

There is a bit of a mismatch because right now because IIUC the search API always returns messages however I can also imagine this filter being useful for message searches. (Ex: messages containing "pizza" in a room that contains Alice and Bob.) Extending the Search API with a `result_type: room` option and adding support for a `member` constraint would add both of these features and provide an obvious way to add more constraints on "Find common rooms" in the future.

Reviewer 6: The `/search` API seems to already have query and result categories (`room_events` being the only currently defined one). It also already supports sender constraints, but that doesn't directly help with room search.

Reviewer 8: That sounds good. So we could add the `rooms` category and add `member` constrains. I agree that `sender` is different than `member`.

Although this does raise fallback questions. If we say that `member` can be used for `room_events` is it required? Or can the server ignore it. But is a server allowed to ignore it for `rooms`? If it does ignore it it makes it useless for the use case described by this MSC.

Reviewer 2: While i think these are good ideas, they kinda fall out of scope of this MSC, the search api needs better tuning in general, and i don't think that bolting something like this onto it is a good idea for now. We should probably revisit this bit when we overhaul the search API in a more broader sense.

Author: +1 I'd rather not encumber the search API with more things when it's already not fit for purpose, and would rather we wrote a MSC specifically to alter how search works separately.

Reviewer 7: Off the back of the above conversation, are there any use cases for allowing clients to hand multiple user IDs to this API, that act as an AND condition to search mutual rooms with?

All I can think of is the use case that @kevincox mentioned above, where you're trying to find a room given some known participants - which actually sounds like a pretty cool feature. We would need to reshape this API to use query parameters instead of path parameters or similar though. Do others think that may be worthwhile, or can think of other use cases?

It's also perfectly valid to just assume that specific use case will be handled by a more featureful room search API.

Reviewer 2: I think that at this point, it's too much of a diversion from what has already settled in over the course of this MSC's development; personally, i'd like to see that in a new endpoint, or in that room search api you mentioned.

(FWIW, the above is talking about a message search API as well, i'm mentioning this to make sure you're not confusing the above two.)

Reviewer 7: I appreciate that this MSC has been open for a while, but I'm not sure if it's _actually_ near-complete. We've yet to reach proposed FCP or have any implementation in use other than [a trivially mutable function](https://github.com/matrix-org/matrix-js-sdk/pull/1417/files#diff-627fe08d8bba6a4fa76e7c5ed36ef6f16d126733e65e19236b391de0d34f1fe0R4852-R4864) written for it. I don't think potentially major changes to the MSC at this point are off the table?

I know that extending this MSC's lifecycle isn't ideal, but I do want to make sure we have a maintainable API.

My current thoughts are:

* If we *do* end up wanting this functionality (**which is not yet clear!**) then we'd be painting ourselves into a corner with the current shape.
* The convention of putting user IDs in path parameters was originally intended to allow AS' to insert puppeted Matrix User ID. But that ended up being scrapped for [the `user_id` query parameter](https://spec.matrix.org/v1.3/application-service-api/#identity-assertion). As a result, ideally we'd remove about half of the instances of user IDs in path parameters; thus it may not be the best precedence to adhere to.

> (FWIW, the above is talking about a message search API as well, i'm mentioning this to make sure you're not confusing the above two.)

Ah indeed it is talking about message search. My comment probably should have been its own thread.

Reviewer 2: We discussed in DMs what the best way forward for this MSC would be, and we came to the conclusions;
- To swap the path parameter to a query parameter.
- To have such an endpoint implicitly include the self-user in the search.
- A possible moderation use-case for this can best be addressed in an admin api.
- If, in the future, a use case comes along where more-than-one users have to be queried, then this endpoint can be easily edited by another MSC, and clients will have to version-check if the server supports multiple users with the query part.

Reviewer 2: /res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

A client which holds full and current state can already see all membership for all rooms, and thus
determine which of those rooms contains a "join" membership for the given user_id. Clients which "lazy-load"
however do not have this information, as they will have only synced a subset of the full membership for
all rooms. While a client *could* pull all membership for all rooms at the point of needing this information,
it's computationally expensive for both the homeserver and the client, as well as a bandwidth waste for constrained
clients.

## Security considerations

The information provided in this endpoint is already accessible to the client if it has a copy of all
state that the user can see. This endpoint only makes it possible to get this information without having
to request all state ahead of time.

## Unstable prefix

The implementation MUST use `/_matrix/client/unstable/uk.half-shot.msc2666/user/mutual_rooms/{user_id}`.

> The /versions endpoint MUST include a new key in `unstable_features` with the name `uk.half-shot.msc2666`.

Reviewer 2: ```suggestion
The /versions endpoint MUST include a new key in `unstable_features` with the name `uk.half-shot.msc2666.mutual_rooms`.
```

See [this implementation review comment](https://github.com/matrix-org/matrix-js-sdk/pull/2271#discussion_r841389852), considering we now have two endpoints for one unstable MSC, we should differentiate the two as well.

Author: Ah, I see. Yes, this seems valid if we've changed the endpoint between versions.

If the value is false or the key is not present, clients MUST assume the feature is not available.

> Once the MSC has been merged, clients should use `/_matrix/client/r0/user/mutual_rooms/{user_id}`

Reviewer 7: ```suggestion
Once the MSC has been merged, clients should use `/_matrix/client/v1/user/mutual_rooms/{user_id}`
```

> Once the MSC has been merged, clients should use `/_matrix/client/v1/user/mutual_rooms/{user_id}`

Reviewer 7: ```suggestion
Once the MSC has been merged, and the homeserver has advertised support for the Matrix version that this endpoint is included in, clients should use `/_matrix/client/v1/user/mutual_rooms/{user_id}`
```

Reviewer 2: 👍

@Half-Shot Could you add this?

Reviewer 2: /res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

such as Telegram offer users the ability to show "groups in common", which allows users to determine
what they have in common before participating in conversation.

There are a variety of applications for this information. Some users may want to block invites from
users they do not share a room with at the client level, and need a way to poll the homeserver for
this information. Another use case would be trying to determine how a user came across your MXID, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,
modern clients often implement [lazy-loading of room members](https://spec.matrix.org/v1.3/client-server-api/#lazy-loading-room-members),
so they often only have a subset of state for the rooms the user is in. Therefore, the homeserver
should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another user.

## Proposal

> Homeservers should implement a new endpoint `/_matrix/client/v1/user/mutual_rooms/{user_id}` which will take

Reviewer 3: ```suggestion
Homeservers implement a new endpoint `/_matrix/client/v1/user/mutual_rooms/{user_id}` which will take
```

"should" implies it's optional when it's not :P

Reviewer 2: /res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

> Homeservers should implement a new endpoint `/_matrix/client/v1/user/mutual_rooms/{user_id}` which will take

Reviewer 3: can it be rate limited?

Reviewer 2: Yes, that should be added. 👍

Reviewer 2: /res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

the MxID of the user who is being searched for.

The response format will be an array containing all rooms where both the authenticated user and `user_id` have
a membership of type `join`. If the `user_id` does not exist, or does not share any rooms with the authenticated user,

> an empty array should be returned.

Reviewer 7: With room types now merged to the spec, should we add a query parameter to constrain the results by type of room? I could see this being useful for:

1. Being able to show a separate list of "Mutual Spaces" and "Mutual rooms".
1. Hiding rooms from the results that shouldn't be shown in clients (i.e. profile rooms).

I also recognise that the client could probably do local filtering, but I'm not sure how different that would be to implement.

Reviewer 2: Personally, I'd like this to be implemented in the Room Search API you alluded to [here](https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r935730582), in a seperate MSC, rather than tacking it onto this one.

I think going forward, local filtering and matching could better address this usecase than not, I think going forward with this MSC as-is is a better option than to wait a longer while for a more fruitful Room Search API.

```
GET /_matrix/client/v1/user/mutual_rooms/%40bob%3Aexample.com
```

```json
{
  "joined": [

>     "!OGEhHVWSdvArJzumhm:matrix.org",

Reviewer 3: Should we be returning stripped state for these rooms too? (or do we plan to do something similar in a future release? if so, making this an object with an empty object as the value would give extensibility).

Reviewer 2: The idea of this endpoint is that it just gives pointers to rooms for an existing "spun-up" client, one which already would have information about these rooms.

However, I don't know the full details of sync v3, and how it'll handle this, will it effectively hide rooms "under the fold", including any and all detail?

If so, then I personally think there's indeed merit in making this future-proof (and thus making it a mapping to empty objects).

However still, if this is going to be the case, I have to think about pagination, which complicates this endpoint a lot. (See your previous comment and my response)

    "!HYlSnuBHTxUPgyZPKC:half-shot.uk",
    "!DueayyFpVTeVOQiYjR:example.com"
  ]
}
```

The response error for trying to get shared rooms with yourself will be an HTTP code 400, with `M_FORBIDDEN` as the `errcode`.

## Potential issues

Homeserver performance and storage may be impacted by this endpoint. While a homeserver already stores
membership information for each of its users, the information may not be stored in a way that is readily

> accessible. Homeservers that have implemented [POST /user_directory/search](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3user_directorysearch)

Reviewer 14: I think the performance impact will be OK if this endpoint is used "infrequently", i.e. like for the use cases described in the intro.

Basically: the query requires looking up the list of rooms for each user and doing the intersection, which is not cheap, but is also not particularly expensive.

Reviewer 2: Yeah, in Element's case, I expect this endpoint to be hit in a few cases;
- Explicitly opening a detail profile view (e.g. clicking on a profile in the sidebar in cinny in element).
- Upon viewing invites (to cross-reference where the person is known from).
- When viewing a minimal DM, where the header ("This is your conversation with X") is visible.

I don't expect it to be pre-loaded en-masse, or loaded en-masse in general.

may have started some of this work, if they are limiting users to searching for users for which they
share rooms. While this is not a given by any means, it may mean that implementations of this API

> and /search may be complimentary.

Reviewer 3: something we've been moving away from is pulling large chunks of data out of the server - a question to consider is if this endpoint needs pagination, or if it's unlikely to be a problem. 

Server-side folks probably have a stronger opinion on this than I.

Reviewer 2: I feel like this is not particularly "large", it's going to be a subset of rooms the user is in, and only then, just their IDs, which would be a string that's reasonably usually around 16 characters.

Even if the user would share around 100 rooms with the other person (a result on the high end, in my opinion), with 16 characters average, it'd only be around ~1.8kb to share (taking into account quotes and commas), which would still be less than a random average (synapse) sync response of around 2KB, and only double of its empty sync response (of ~700B).

Even if the numbers start to rise (beyond 1000 shared rooms), I don't think this'll ever get into the Megabyte territory of responses, and thus require pagination.

Thoughts?

Reviewer 14: I'd quite like us to move to a world where every endpoint that responds with a variable amount of data is paginated, for a few reasons:

1. it's much less faff for clients if the endpoints start off being paginated compared with introducing pagination later;
2. if everything is paginated and follow a similar structure then its just easier for clients and servers to deal with; and
3. even if the common case only returns a small amount of data, the mere potential that a client *can* cause lots of data to be returned makes it hard to run Synapse in a resource constrained way.

Also, if we end up extending the API to include stripped state then the size of the response can easily balloon.

Reviewer 2: /res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

We didn't need to specify the order the rooms are returned in, but allow the server to specify the ordering itself, and simply return an opaque token to get the next batch of rooms

this'd allow it to define an ordering itself, with any kind of optimisation strategy

this information. Another use case would be trying to determine how a user came across your MXID, as
invites on their own do not present much context. With this endpoint, a client could tell you what
rooms you have in common before you accept an invite.

While this information can be determined if the user has full access to member state for all rooms,
modern clients often implement "lazy-loading", so they often only have a subset of state for the rooms
the user is in. Therefore, the homeserver should have a means to provide this information.

This proposal aims to implement a simple mechanism to fetch rooms you have in common with another user.

## Proposal

Homeservers should implement a new endpoint `/user/mutual_rooms/{user_id}` which will take
the MxID of the user who is being searched for.

The response format will be an array containing all rooms where both the authenticated user and `user_id` have
a membership of type `join`. If the `user_id` does not exist, or does not share any rooms with the authenticated user,
an empty array should be returned.

Additionally, a server may respond with a `400` `M_UNKNOWN` to this endpoint for any reason, 

> signalling its unwillingess (or inability) to satisfy the request.

Reviewer 3: this is a very uncommon thing to do in the spec. All endpoints are currently required to be implemented (unless bundled into a module the server doesn't care about), so this endpoint should never fall into "unwillingness" or "inability". 

If we do want to do this, please provide justification in the MSC.

Reviewer 2: I'm conflicted on this requirement myself also, I'm starting to think that the arbitrary requirement from Synapse - where this is implemented via the optionally-ran user directory - would not be enough to restrict this MSC to have to include this.

I will take another look at the Synapse implementation, see if it can be implemented in a way that makes this section redundant.

Reviewer 3: Conveniently, the Synapse implementation is what drove this thread being opened. Implementation-specific restrictions are not supposed to be carried into the protocol (but cases where many/all implementations have issues would be considered).

For this case, the Synapse side is kinda stuck at "figure it out" rather than having the spec concede.

Reviewer 2: https://github.com/matrix-org/synapse/pull/12836 pushed a change to synapse, I'll keep this open until that has been resolved.

Reviewer 2: This has been resolved on both fronts now.

Reviewer 2: https://github.com/matrix-org/matrix-spec-proposals/pull/3822 has also been made to address the proposal-text-side of things.

Reviewer 2: /res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913
